craft_parts/permissions.py

```diff
@@ -0,0 +1,102 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2022 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Specify and apply permissions and ownership to part-owned files."""
+
+import os
+from fnmatch import fnmatch
+from pathlib import Path
+from typing import List, Optional, Union
+
+from pydantic import BaseModel, root_validator
+
+
+class Permissions(BaseModel):
+    """Description of the ownership and permission settings for a set of files.
+
+    A `Permissions` object specifies that a given pattern-like `path` should
+    be owned by `owner` with a given `group`, and have the read/write/execute
+    bits defined by `mode`.
+
+    Notes
+    -----
+      - `path` is optional and defaults to "everything";
+      - `owner` and `group` are optional if both are omitted - that is, if
+        one of the pair is specified then both must be;
+      - `mode` is a string containing an integer in base 8. For example, "755",
+      "0755" and "0o755" are all accepted and are the equivalent of calling
+      `chmod 755 ...`.
+
+    """
+
+    path: str = "*"
```

Comment:
Should the default path be `*`, or `**/*`? The latter case would match every file regardless of directory.

Reply:
So the reason that I used `"*"` is because I'm using `fnmatch()` instead of, for example, `glob()`. `fnmatch()` works with strings as a "simplified" regex, so `"*"` matches *everything*.

Now, the reason that I'm using `fnmatch()` is because I'm thinking that this will be integrated into `migrate_files()`, which already receives the full set of files/paths to copy. So, while `glob()` takes a pattern and gives you the filenames that would match it, I kind of need the opposite: given a filepath, does this pattern match it? That's why I used `fnmatch()`.

Do you have a good idea for a better way to do this? How do we do this elsewhere?

Reply:
Ah, I was unaware that fnmatch would match it that way. Since just `*` sets the default behaviour we want it should be fine.